src/nexusGrades.js

```diff
@@ -198,60 +247,65 @@ fluid.defaults("fluid.nexus.bindModel.handler", {
+            funcName: "fluid.nexus.bindModel.removeListenerOnDestroy",
+            args: ["{that}"]
         }
     }
 });
 
-fluid.nexus.bindModel.bindWs = function (handler, componentPath, modelPath, modelChangeListener, componentRoot) {
-    handler.componentHolder.targetComponent = fluid.nexus.componentForPathInContainer(componentRoot, componentPath);
-    // TODO: Note that applier.modelchanged.addListener is different from https://wiki.fluidproject.org/display/fluid/Nexus+API
-    //       Which says applier.addModelListener
-    // if the modelPath is undefined, the binding is to the entire model
-    modelPath = modelPath || "";
-    handler.modelPathSegs = fluid.pathUtil.parseEL(modelPath);
-    handler.targetModelChangeListenerId = fluid.allocateGuid();
-    handler.componentHolder.targetComponent.applier.modelChanged.addListener(
+fluid.nexus.bindModel.bindWs = function (that, modelChangeListener) {
+    that.targetModelChangeListenerId = fluid.allocateGuid();
+    that.componentHolder.targetComponent.applier.modelChanged.addListener(
         {
-            segs: handler.modelPathSegs,
-            listenerId: handler.targetModelChangeListenerId
+            segs: that.modelPathSegs,
+            listenerId: that.targetModelChangeListenerId
         },
         modelChangeListener
     ); // TODO: namespace?
 
     // On connect, send a message with the state of the component's model at modelPath
-    handler.sendMessage(fluid.get(handler.componentHolder.targetComponent.model, handler.modelPathSegs));
+    that.sendTypedMessage("initModel", fluid.get(that.componentHolder.targetComponent.model, that.modelPathSegs));
 };
 
-fluid.nexus.bindModel.targetModelChangeListener = function (handler, value) {
-    handler.sendMessage(value);
+fluid.nexus.bindModel.targetModelChangeListener = function (that, value) {
+    that.sendTypedMessage("modelChanged", value);
 };
 
-fluid.nexus.bindModel.receiveMessage = function (component, baseModelPathSegs, message) {
+fluid.nexus.bindModel.receiveMessage = function (that, targetComponent, baseModelPathSegs, message) {
+    if (message.path === undefined || message.value === undefined) {
+        that.sendTypedMessage("error", fluid.get(targetComponent.model, baseModelPathSegs));
+        return;
+    };
     var messagePathSegs = fluid.pathUtil.parseEL(message.path);
     var changePathSegs = baseModelPathSegs.concat(messagePathSegs);
-    component.applier.fireChangeRequest(
+    targetComponent.applier.fireChangeRequest(
         {
             segs: changePathSegs,
             value: message.value,
             type: message.type
         }
     );
 };
+
+// This guarded listener destruction is needed because the bindModel websocket component
+// may be destroyed early if ta 404 response is emitted during the HTTTP setup,
+// in which case the listener to be removed has not been mounted yet.
+// FIXME: I think this should be a Kettle issue
```

Comment:
I think not necessarily, if you have conditionally added a listener you will need to conditionally remove it as well. Naturally it would be better if there were a declarative pattern for this - we might be able to throw something together using options distributions but it is probably not worth it in what is so naturally patterned as rather intricate integration code. If there is a fix anywhere, it might be in core Infusion if we find that the modelChanged listener cannot be automatically deregistered if the component distributing it down is destroyed. But I don't believe that core Infusion is currently capable of applying options distributions across injection boundaries anyway - this is a miracle that you only recently convinced me might be possible at all (permeable transclusion boundaries).

Reply:
To further elaborate on this point - the real barrier is that the framework is currently not capable of performing an asynchronous action during component construction, unless that action is expressed as a "resource" attached to a ResourceLoader. This is a conceivable, though slightly unidiomatic refactoring of the middleware component - that is, we could convert the "handle" method into a promise-backed resource, but this still doesn't solve the conundrum of how we end up with what morally speaking is an injected subcomponent which arrives asynchronously.

We were not able as part of the FLUID-6148 rewrite to enable "general asynchrony" - that is, to allow absolutely any component option or other material to arrive asynchronously as the headline JIRA originally described, since this would have degraded the already very poor framework performance unacceptably as every inner loop in the options merging pathway would need to be opened out and check whether a value delivered from an expander was a promise or not.

However, two weeks ago a little after we were speaking about this, I had the SOFA MORNING INSIGHT that we could combine the https://issues.fluidproject.org/browse/FLUID-5304 "Infusion Compiler" work with the rest of the FLUID-6148 work by means of implementing what could be called ASYNCHRONOUS SPILL. The insight recognises that "being asynchronous" could be just another form of value polymorphism recognised by the compiler, causing it to spill its cache of monomorphised grade definitions if its expectations accumulated during a previous expansion were violated. That is, asynchrony polymorphism (discovering that a site was asynchronous where previously found to be synchronous or vice versa) would be on equal terms with the "value shape polymorphism" where an expanded value was found to be of a JSONically different shape to its predecessor in a way that disrupted the merge compositing algorithm. This would all be a form of "symbolic execution" as it is termed in the CS literature.